Improve Daml installation documentation for "No space left" error

cristina-da opened the pull request:
Fixes #21348

<!--
# Pull Request Checklist

- Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- Include appropriate tests
- Set a descriptive title and thorough description
- Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- Normal production system change, include purpose of change in description
- If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.
-->

cristina-da commented on sdk/docs/manually-written/sdk/tutorials/setup-assistant/installation.rst:
```diff
@@ -136,6 +136,28 @@ Download Manually
+Troubleshooting
+***************
+
+The Daml installation requires x amount of disk space. If there is not enough space on disk, the installer outputs the error:
```
"x" here should be replaced with some measure of the total space required, perhaps the reviewers have an idea?

meiersi-da replied:
I've had a quick look at the SDK installations I have on my machine:
2.10 -> 1.6 GB
3.3.0 -> 1.1 GB

I'd suggest to write `around 1.5 GB` of disk space.

meiersi-da commented on sdk/docs/manually-written/sdk/tutorials/setup-assistant/installation.rst:
```diff
@@ -136,6 +136,28 @@ Download Manually
+
+If the installer cannot ensure sufficient space in the the user's home directory and the temporary files directory, an alternative is to change
+the location of these directories before the installation.
+To do so, please set the ``DAML_HOME`` and ``TMPDIR`` environment variables to other directories that have sufficient space. You can find instructions on how to do this :doc:`here <path-variables>`.
```
please avoid "here" references as per style-guide: https://docs.google.com/document/d/153D6-gI0blsAvlrzReHVyz8RCHUQzzdKnxnpcADlt6c/edit?tab=t.0#bookmark=id.ya2a6pwloo5b. Also a `:ref:` is preferred, as it is more stable under doc refactorings than a `:doc:` reference.

meiersi-da replied:
Any chance to inline the command line:
```
DAML_HOME=<YOUR-TARGET-DIR> TMPDIR=<YOUR-TEMPDIR> curl -sSL https://get.daml.com/ | sh
```

meiersi-da commented on sdk/docs/manually-written/sdk/tutorials/setup-assistant/installation.rst:
```diff
@@ -136,6 +136,28 @@ Download Manually
+
+If the installer cannot ensure sufficient space in the the user's home directory and the temporary files directory, an alternative is to change
+the location of these directories before the installation.
+To do so, please set the ``DAML_HOME`` and ``TMPDIR`` environment variables to other directories that have sufficient space. You can find instructions on how to do this :doc:`here <path-variables>`.
```
```suggestion
To do so, call the installer with ``DAML_HOME`` and ``TMPDIR`` environment variables pointing to directories that have sufficient space:

.. code-block: shell
  
  DAML_HOME=<YOUR-TARGET-DIR> TMPDIR=<YOUR-TEMPDIR> curl -sSL https://get.daml.com/ | sh
```

(Not sure about the code-block syntax.)

cristina-da replied:
Done. I've added slightly different instructions for Windows, which should be fine, but could you please take another look?

meiersi-da commented on sdk/docs/manually-written/sdk/tutorials/setup-assistant/installation.rst:
```diff
@@ -136,6 +136,28 @@ Download Manually
+If the installer cannot ensure sufficient space in the the user's home directory and the temporary files directory, an alternative is to change
+the location of these directories before the installation.
+To do so, please set the ``DAML_HOME`` and ``TMPDIR`` environment variables to other directories that have sufficient space. You can find instructions on how to do this :doc:`here <path-variables>`.
+It is important that the ``DAML_HOME`` and ``TMPDIR`` environment variables do not point to directories located inside the unzipped installation
```
```suggestion
Avoid choosing directories for ``DAML_HOME`` and ``TMPDIR`` environment variables that are located inside the unzipped installation
```

meiersi-da commented on sdk/docs/manually-written/sdk/tutorials/setup-assistant/installation.rst:
```diff
@@ -136,6 +136,28 @@ Download Manually
+the location of these directories before the installation.
+To do so, please set the ``DAML_HOME`` and ``TMPDIR`` environment variables to other directories that have sufficient space. You can find instructions on how to do this :doc:`here <path-variables>`.
+It is important that the ``DAML_HOME`` and ``TMPDIR`` environment variables do not point to directories located inside the unzipped installation
+direcory ``sdk-x.x.x``. If they do, the installation does not complete.
```
```suggestion
direcory ``sdk-x.x.x``, as otherwise the installer fails.
```

meiersi-da reviewed (changes requested):
Please use the `release/3.3.x` branch as the target. Docs changes currently go there while we haven't ported all of the changes from 3.3 to `main`

cristina-da commented:
> Please use the `release/3.3.x` branch as the target. Docs changes currently go there while we haven't ported all of the changes from 3.3 to `main`

I couldn't find the installation page on `release/3.3.x`, even after grepping for contents that would likely be on that page. Could you point me to it please?

meiersi-da commented:
> > Please use the `release/3.3.x` branch as the target. Docs changes currently go there while we haven't ported all of the changes from 3.3 to `main`
> 
> @meiersi-da I couldn't find the installation page on `release/3.3.x`, even after grepping for contents that would likely be on that page. Could you point me to it please?

It's here: https://docs.digitalasset-staging.com/build/3.3/component-howtos/smart-contracts/assistant#install

I'd add your content as a separate subsection to https://docs.digitalasset-staging.com/build/3.3/component-howtos/smart-contracts/assistant#troubleshoot

remyhaemmerle-da commented:
/azp run

dylant-da commented on sdk/docs/manually-written/sdk/tutorials/setup-assistant/installation.rst:
```diff
@@ -14,7 +14,7 @@ You need to install:
 2. JDK 11 or greater. If you don't already have a JDK installed, try `Eclipse Adoptium <https://adoptium.net>`_.
 
-   As part of the installation process you may need to set up the ``JAVA_HOME`` variable. You can find instructions for this process on :doc:`Windows,macOS, and Linux here <path-variables>`.
+   As part of the installation process you may need to set up the ``JAVA_HOME`` variable. You can find instructions for this process on :doc:`Windows,macOS, and Linux in the :ref:`path-variables <path-variables>` section.
```
I don't think the `ref` is necessary here, since we're already inside of a :doc: directive this probably breaks compilation.

dylant-da commented on sdk/docs/manually-written/sdk/tutorials/setup-assistant/installation.rst:
```diff
@@ -136,6 +136,36 @@ Download Manually
+
+* On Windows, one can find the user's home directory in the `%USERPROFILE%` environment variable, and the temporary files directory in the
+  ``%TEMP%`` environment variable
+* On Linux and MacOS, one can find the user's home directory in the ``HOME%`` environment variable, and the temporary files directory in the
```
I don't think that `%` symbol is accurate, no?
```suggestion
* On Linux and MacOS, one can find the user's home directory in the ``HOME`` environment variable, and the temporary files directory in the
```

dylant-da commented on sdk/docs/manually-written/sdk/tutorials/setup-assistant/installation.rst:
```diff
@@ -136,6 +136,36 @@ Download Manually
+In order to fix this problem, one needs to ensure sufficient disk space overall, and in particuloar that the user's home directory and the temporary
+files directory have sufficient space (i.e., in the order of 2GiB each):
+
+* On Windows, one can find the user's home directory in the `%USERPROFILE%` environment variable, and the temporary files directory in the
```
I think we need to mention DAML_HOME instead of HOME here, since HOME is only partially relevant for where the SDK installs.

According to our code:

```
-- | The DAML_HOME environment variable determines the path of the daml
-- assistant data directory. This defaults to:
--
--     System.Directory.getAppUserDataDirectory "daml"
--
-- On Linux and Mac, that's ~/.daml
-- On Windows, that's %APPDATA%/daml
```

So we could tell the user that DAML_HOME sets where Daml SDK is installed, and that it defaults to the locations listed above.

dylant-da commented on sdk/docs/manually-written/sdk/tutorials/setup-assistant/installation.rst:
```diff
@@ -136,6 +136,36 @@ Download Manually
+* On Linux and MacOS, one can find the user's home directory in the ``HOME%`` environment variable, and the temporary files directory in the
+  ``TMPDIR`` environment variable 
+
+If the installer cannot ensure sufficient space in the the user's home directory and the temporary files directory, an alternative is to change
```
```suggestion
If the installer cannot ensure sufficient space in the Daml SDK's home directory (``DAML_HOME``) and the temporary files directory, an alternative is to change
```

Depending on how you word the section before this one, just saying the name of the environment variable may be more appropriate.

dylant-da commented on sdk/docs/manually-written/sdk/tutorials/setup-assistant/installation.rst:
```diff
@@ -136,6 +136,36 @@ Download Manually
+
+Avoid choosing directories for ``DAML_HOME`` and ``TMPDIR`` environment variables that are located inside the unzipped installation
+directory ``sdk-x.x.x``, as otherwise the installer fails.
+
```
Recommending DAML_HOME also requires that we update the docs for path-variables.rst, since there we tell people to just use commands like
```
        echo 'export PATH="$HOME/.daml/bin:$PATH"' >> ~/.bash_profile
```
which won't be correct when DAML_HOME is set.

dylant-da commented on sdk/docs/manually-written/sdk/tutorials/setup-assistant/installation.rst:
```diff
@@ -136,6 +136,36 @@ Download Manually
+ 
+.. code-block: bash
+  
+  DAML_HOME=<YOUR-TARGET-DIR> TMPDIR=<YOUR-TEMPDIR> curl -sSL https://get.daml.com/ | sh
```
We should also mention that you'll need to have the same DAML_HOME set and add `$DAML_HOME/bin` to the path in all terminals you use in the future, probably in .bashrc. Up to you if you want to include this here or in path-variables.rst

dylant-da reviewed (commented):
This is definitely an underdocumented part of the SDK, so thank you for taking on the task of fixing it. Looks pretty good so far, but I've reviewed a few changes that I think are needed to make sure clients end up with a working PATH and installation.

meiersi-da commented:
> @meiersi-da Thank you for the quick review! I've made the changes, could you please take another look when you've got time?

I took another look. Looks good content-wise, assuming you address the comments by @dylant-da. It seems though that you are still targeting the `main` branch and not `release/3.3.x`, correct?

cristina-da commented:
@dylant-da Thanks a lot for the review! I've addressed your comments, please take a look. FYI, I'll rebase on `release/3.3.x` in my next commit.

cristina-da commented on sdk/docs/manually-written/sdk/component-howtos/smart-contracts/assistant.rst:
```diff
@@ -28,7 +28,7 @@ These methods will install the most recent Daml SDK with its Daml Assistant, and
 *********
 Daml assistant uses a ``daml-config.yaml`` file to configure SDK installation behaviour.  
-On Linux/MacOS, this will be in ``~/.daml/``. On Windows, it will be in ``C:/Users/<user>/AppData/Roaming/daml``.  
+On Linux/MacOS, this will be in ``$DAML_HOME``. On Windows, it will be in ``%DAML_HOME%``.  
```
And same for the changes below (before troubleshooting). Thank you!

cristina-da commented:
@meiersi-da I saw that, in the meantime, you ported all release-line changes to main https://daholdings.slack.com/archives/CEP4Q823G/p1749743408898609. Should I then port these (once merged) to main? Thanks.

dylant-da commented on sdk/docs/manually-written/sdk/component-howtos/smart-contracts/assistant.rst:
```diff
@@ -143,12 +143,40 @@ Troubleshoot
+an alternative is to change the location of these directories before the installation.
+To do so, call the installer with ``DAML_HOME`` and ``TMPDIR`` environment variables pointing to directories that have sufficient space:
+
+* On Windows, first set the environment variables. Then run the installation as usual.
```
I think this suffices - I haven't enough experience with Windows to expand upon this myself.

dylant-da reviewed (approved):
Looks good, approved - I'm guessing the path-variables changes will have to go in a separate PR to docs.daml.com, if you ping me I can stamp it. Thank you for documenting this!

cristina-da commented:
I have hereby read the Digital Asset CLA and agree to its terms

cristina-da commented:
@meiersi-da Have my changes (including change if target to `release/3.3.x`) addressed your feedback? I'm asking because, even though you agreed in a comment with the changes, I still see `Requested changes` from your review, and that is blocking the merge.
